### PR TITLE
Jetpack Migration: Do not direct to WordPress if there are prior migration errors

### PR DESCRIPTION
**Pull request by dvdchr**
Refs: pcdRpT-1sp-p2#comment-2466, pcdRpT-1sp-p2#comment-2478

As titled, this PR stores the error produced from `ContentMigrationCoordinator` to the App Group. The purpose is for Jetpack to use this information to determine if it should present the sign-in flow or the "Open WordPress" flow.

> **Note**: As a reminder, we show the Load WordPress flow for users that opens the Jetpack app while having the WordPress but have never opened it at least once (to trigger the export process).

The "Open WordPress" flow is now shown when:
- The import process failed,
- The user has a compatible WordPress app installed,
- There is no exported data in the App Group space,
- The WordPress app has no previous export failures.

To enter this flow, we want to be sure that **there is no data to import because the user has never triggered the export process and not because the WP app has previously tried and failed.** Any condition other than the above should bring the user to the sign-in flow.

## To test

As a reminder, here are three possible states when the Jetpack app is opened:

🔐 Sign-in flow | 🟢 Normal migration | 🔁 Open WordPress
-|-|-
Splash | Welcome | Welcome - JP entry point

Next, refer to the table below to see possible setups and variations, along with the expected flow when Jetpack is opened. To test, please follow the conditions in columns 2-4, and verify that the expected flow in the last column is correct.

No. | WordPress app | Account state | Scenario | Expected flow in Jetpack
-|-|-|-|-
1 | Not installed | n/a | n/a | 🔐 Sign-in flow
2 | Incompatible[^4] | Any | Any | 🔐 Sign-in flow
3 | Compatible | Logged out | Never opened | 🔁 Open WordPress
4 | Compatible | Logged out | Opened at least once | 🔐 Sign-in flow[^2]
5 | Compatible | Logged out | Log in > then tap the Jetpack banner | 🟢 Normal migration
6 | Compatible | Logged in | Never opened[^1] | 🔁 Open WordPress
7 | Compatible | Logged in | Opened at least once | 🟢 Normal migration
8 | Compatible | Logged in | Tap the Jetpack banner | 🟢 Normal migration
9 | Compatible | Logged in | Airplane mode > create local draft > tap the Jetpack banner | 🔐 Sign-in flow[^2]
10 | Compatible | Logged in | Log out | 🔐 Sign-in flow[^2]

To be clear, this PR updates the expected flow from **🔁 Open WordPress** to **🔐 Sign-in flow** for rows 4, 9, and 10. However, I'm listing all the other pathways (that I can think of) to ensure they're not impacted by this change.

---

💡 **Pro tip:** If you're planning to test by compiling directly to your device or simulator, it can be tricky to test the `Never opened` scenario. Here's how you can configure Xcode to not launch the app immediately after it finishes building:

1. Click on the app scheme (WordPress or Jetpack), then Edit Scheme...
2. Select the Run tab.
3. In the Launch section, select the "Wait for the executable to be launched" option.

---

## Regression Notes
1. Potential unintended areas of impact
Some migration entry points may be affected. Let's hope not 😬 .

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes.

3. What automated tests I added (or what prevented me from doing so)
Added unit tests to cover additional logic in `ContentMigrationCoordinator`.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

[^1]: This is an app update case. From a previous WordPress app, log in to any account, and update the app to the latest version. But DO NOT open it after updating.
[^2]: Previously, this would be 🔁 Open WordPress.
[^4]: This would be version 21.2 and below.

### CODE REVIEW

File: WordPress/Jetpack/Classes/Utility/DataMigrator.swift
Comment (dvdchr) [inline WordPress/Jetpack/Classes/Utility/DataMigrator.swift:23]: cc: @salimbraksa – this error is used from both the WP and the JP side. So I think this change should cover both scenarios. 🙂

Comment (salimbraksa) [reply WordPress/Jetpack/Classes/Utility/DataMigrator.swift:23]: @dvdchr Yes, thanks for including this tweak in this PR 🙇

File: WordPress/Classes/Utility/Migration/ContentMigrationCoordinator.swift
Comment (salimbraksa) [inline WordPress/Classes/Utility/Migration/ContentMigrationCoordinator.swift]: I think if we want to present an Alert ( or Error Screen ) on the Jetpack side, we should store the `rawValue` of `ContentMigrationCoordinatorError` instead of `errorDescription`, then we let the UI side to decide how to format the error and display it to the user. What do you think?

Comment (dvdchr) [reply WordPress/Classes/Utility/Migration/ContentMigrationCoordinator.swift]: Good point. This is now addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/19787/commits/95a1ccc320af53cbc34a2689d68fcf1868b4858b!

File: WordPress/Classes/Utility/Migration/ContentMigrationCoordinator.swift
Comment (twstokes) [inline WordPress/Classes/Utility/Migration/ContentMigrationCoordinator.swift]: Non-blocking nit / preference: 

A singular `Error` seems like a better fit for this, but could introduce extra confusion with [the protocol name](https://developer.apple.com/documentation/swift/error). `MigrationError` may be a good fit for its purpose and reduces the nested type stuttering (e.g. `ContentMigrationCoordinator.ContentMigrationCoordinatorError`), but that's just my opinion. 😄

Comment (dvdchr) [reply WordPress/Classes/Utility/Migration/ContentMigrationCoordinator.swift]: I don't have any preference, but I definitely don't want a Java-esque `ContentMigrationCoordinator.ContentMigrationCoordinatorError` 😂 . Anywho, addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/19787/commits/24e403aba234a86ebee1e96343d8727b9155cc78.

File: WordPress/Classes/Utility/Migration/ContentMigrationCoordinator.swift
Comment (twstokes) [inline WordPress/Classes/Utility/Migration/ContentMigrationCoordinator.swift:72]: 👋 @dvdchr - it looks like Jetpack also runs this sequence and would write out an `.ineligible` failure at least once - do we want that? It [gets called from here](https://github.com/wordpress-mobile/WordPress-iOS/blob/24e403aba234a86ebee1e96343d8727b9155cc78/WordPress/Classes/System/WordPressAppDelegate.swift#L332).

I discovered this because I was testing `21.4` Jetpack with `21.3` WordPress (though I see your footnote about `21.2`, so perhaps this isn't a possible combo) and we're getting `true` for `isCompatibleWordPressAppPresent` since we only disabled the feature flags but not the URL schemes. 😬 

I'd appreciate another set of eyes on the problem, but I wanted to note two potential changes that will be semi-related to this PR:
- Adding a `if AppConfiguration.isWordPress` somewhere for the `startOnceIfNeeded` sequence so that it only runs in WordPress
- Removing the `wordpressmigration+v1://` URL scheme - **Update:** The `21.3` release may be skipped (I'm going to get confirmation on that) and if so, we have no problem here. 👍 

Thanks!

**Comment by twstokes** [comment]
Here are the tests I've been able to run so far:

| Test | Result | Note |
| - | - | - |
| 1 | ✅ |
| 2 | ✅ |
| 3 | ✅ |
| 4 | ✅ |
| 5 | ✅ |
| 6 | ✅ | Works after merging in https://github.com/wordpress-mobile/WordPress-iOS/pull/19792 |
| 7 | ✅ |
| 8 | ✅ |
| 9 | ✅  | Perhaps in a future PR we should give feedback to the user since it isn't clear that the export failed and we aren't going to migrate their data over.  |
| 10 | ✅  |

### Test 6

**Edit:** This now works - see note above. 🥳

I attempted this twice and my steps were:
1. Installed WordPress 21.2 from TestFlight
2. Authenticated with a .com account and exited the app
3. Installed WordPress from the branch and didn't open it (simulating an upgrade)
4. Installed Jetpack from the branch and opened it
5. I saw the 🔁 Open WordPress view as expected
6. Tapping the "Open WordPress" button properly sent me to WordPress and back to Jetpack

- ⚠️ On attempt 1 it just looped between the two apps
- ⚠️ On attempt 2 WordPress crashed with an error:

```
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSArrayM replaceObjectAtIndex:withObject:]: object cannot be nil'
```

**Comment by dvdchr** [comment]
@twstokes thank you for your help! 🙂 

> it looks like Jetpack also runs this sequence and would write out an `.ineligible` failure at least once - do we want that?

Oof, good catch. This was due to a sloppy refactor work in https://github.com/wordpress-mobile/WordPress-iOS/pull/19787/commits/d1cad49ba56a38ded2c8de8bd2e2b9c18d10aa8d. Previously the `processResult` method was a local closure variable inside `startAndDo`. I placed the `AppConfiguration` guard there and tested some cases before moving it into a separate method, but ended up forgetting to re-add the guard condition 😓. This is now resolved in https://github.com/wordpress-mobile/WordPress-iOS/pull/19787/commits/42789939adc085145cd14416742b89845447bf21. 

> Tapping the "Open WordPress" button properly sent me to WordPress and back to Jetpack

I do remember a strange crash during my first round of testing, but I dismissed it because I thought it was due to a loose cable connection. And the fact that the next attempt worked. I've tested scenario 6 again several times, and this time the crash looks much more consistent to me. Tapping "Open WordPress" the first and the second time always crashed WordPress, but somehow the third time _always_ works 🙃 . For now, I'll note down the stack trace below.

<details>
<summary>📜 WordPress crash information</summary>

```
Exception Type:  EXC_CRASH (SIGABRT)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Triggered by Thread:  2

Application Specific Information:
abort() called


Last Exception Backtrace:
0   CoreFoundation                	       0x18b19de88 __exceptionPreprocess + 164
1   libobjc.A.dylib               	       0x1844cb8d8 objc_exception_throw + 60
2   CoreFoundation                	       0x18b3430a8 -[__NSCFString characterAtIndex:].cold.1 + 0
3   CoreFoundation                	       0x18b33fb30 -[__NSArrayM replaceObjectAtIndex:withObject:].cold.2 + 0
4   CoreFoundation                	       0x18b1c66fc -[__NSArrayM replaceObjectAtIndex:withObject:] + 364
5   CoreData                      	       0x192b05fd4 __84-[NSPersistentStoreCoordinator migratePersistentStore:toURL:options:withType:error:]_block_invoke + 1432
6   CoreData                      	       0x192a50840 gutsOfBlockToNSPersistentStoreCoordinatorPerform + 204
7   libdispatch.dylib             	       0x19276dfdc _dispatch_client_callout + 20
8   libdispatch.dylib             	       0x19277d574 _dispatch_lane_barrier_sync_invoke_and_complete + 56
9   CoreData                      	       0x192b045a0 _perform + 176
10  CoreData                      	       0x192b05960 -[NSPersistentStoreCoordinator migratePersistentStore:toURL:options:withType:error:] + 300
11  WordPress                     	       0x100edec1c CoreDataStack.createStoreCopy(to:) + 2456
12  WordPress                     	       0x102307ec4 DataMigrator.copyDatabase(to:) + 140
13  WordPress                     	       0x102307c24 DataMigrator.exportData(completion:) + 352
14  WordPress                     	       0x102309444 protocol witness for ContentDataMigrating.exportData(completion:) in conformance DataMigrator + 20
15  WordPress                     	       0x1010f2edc ContentMigrationCoordinator.startAndDo(completion:) + 560
16  WordPress                     	       0x1010f3f10 ContentMigrationCoordinator.startOnceIfNeeded(completion:) + 528
17  WordPress                     	       0x101ff7d40 closure #2 in WordPressAppDelegate.runStartupSequence(with:) + 140
18  WordPress                     	       0x100ebb1e8 thunk for @escaping @callee_guaranteed () -> () + 48
19  libdispatch.dylib             	       0x19276c4b4 _dispatch_call_block_and_release + 32
20  libdispatch.dylib             	       0x19276dfdc _dispatch_client_callout + 20
21  libdispatch.dylib             	       0x1927710c8 _dispatch_queue_override_invoke + 788
22  libdispatch.dylib             	       0x19277fa6c _dispatch_root_queue_drain + 396
23  libdispatch.dylib             	       0x192780284 _dispatch_worker_thread2 + 164
24  libsystem_pthread.dylib       	       0x1d81f3dbc _pthread_wqthread + 228
25  libsystem_pthread.dylib       	       0x1d81f3b98 start_wqthread + 8
```

</details>

You also mentioned some oddities about logging out from Jetpack after the migration in pcdRpT-1sp-p2#comment-2524 — in my case, the Jetpack app crashed 😱 .

<details>
<summary>📜 Jetpack crash information</summary>

```
Exception Type:  EXC_CRASH (SIGABRT)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Triggered by Thread:  0

Application Specific Information:
abort() called


Last Exception Backtrace:
0   CoreFoundation                	       0x18b19de88 __exceptionPreprocess + 164
1   libobjc.A.dylib               	       0x1844cb8d8 objc_exception_throw + 60
2   CFNetwork                     	       0x18c33c4c8 0x18c316000 + 156872
3   Jetpack                       	       0x1065a7ecc closure #1 in DataRequest.Requestable.task(session:adapter:queue:) + 200
4   Jetpack                       	       0x1065ad538 partial apply for closure #1 in DataRequest.Requestable.task(session:adapter:queue:) + 28
5   libswiftDispatch.dylib        	       0x193947900 partial apply for thunk for @callee_guaranteed () -> (@out A, @error @owned Error) + 28
6   libswiftDispatch.dylib        	       0x19394ec88 thunk for @callee_guaranteed () -> (@out A, @error @owned Error)partial apply + 16
7   libswiftDispatch.dylib        	       0x193948c08 closure #1 in closure #1 in OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 188
8   libswiftDispatch.dylib        	       0x193948d1c partial apply for thunk for @callee_guaranteed () -> () + 28
9   libswiftDispatch.dylib        	       0x1939486e8 thunk for @escaping @callee_guaranteed () -> () + 28
10  libdispatch.dylib             	       0x19276dfdc _dispatch_client_callout + 20
11  libdispatch.dylib             	       0x19277d574 _dispatch_lane_barrier_sync_invoke_and_complete + 56
12  libswiftDispatch.dylib        	       0x193948040 implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 180
13  libswiftDispatch.dylib        	       0x193947060 partial apply for implicit closure #2 in implicit closure #1 in OS_dispatch_queue.sync<A>(execute:) + 56
14  libswiftDispatch.dylib        	       0x193947d50 OS_dispatch_queue._syncHelper<A>(fn:execute:rescue:) + 396
15  libswiftDispatch.dylib        	       0x19394710c OS_dispatch_queue.sync<A>(execute:) + 164
16  Jetpack                       	       0x1065a7ce8 DataRequest.Requestable.task(session:adapter:queue:) + 480
17  Jetpack                       	       0x1065dff74 SessionManager.request(_:) + 1052
18  Jetpack                       	       0x1065dfa08 SessionManager.request(_:method:parameters:encoding:headers:) + 1272
19  Jetpack                       	       0x107419ad4 WordPressComRestApi.request(method:urlString:parameters:encoding:success:failure:) + 684
20  Jetpack                       	       0x10741c158 WordPressComRestApi.GET(_:parameters:success:failure:) + 224
21  Jetpack                       	       0x1072d91ac NotificationSyncServiceRemote.loadNotes(withNoteIds:fields:pageSize:completion:) + 1420
22  Jetpack                       	       0x1072d8b3c NotificationSyncServiceRemote.loadNotes(withPageSize:noteIds:completion:) + 256
23  Jetpack                       	       0x104e40dbc closure #1 in closure #1 in closure #1 in NotificationSyncMediator.sync(_:) + 296
24  Jetpack                       	       0x104e44920 closure #4 in closure #1 in closure #1 in NotificationSyncMediator.determineUpdatedNotes(with:completion:) + 80
25  Jetpack                       	       0x104dd5c38 thunk for @escaping @callee_guaranteed () -> () + 48
26  libdispatch.dylib             	       0x19276c4b4 _dispatch_call_block_and_release + 32
27  libdispatch.dylib             	       0x19276dfdc _dispatch_client_callout + 20
28  libdispatch.dylib             	       0x19277c7f4 _dispatch_main_queue_drain + 928
29  libdispatch.dylib             	       0x19277c444 _dispatch_main_queue_callback_4CF + 44
30  CoreFoundation                	       0x18b22e6f8 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 16
31  CoreFoundation                	       0x18b210058 __CFRunLoopRun + 2036
32  CoreFoundation                	       0x18b214ed4 CFRunLoopRunSpecific + 612
33  GraphicsServices              	       0x1c4512368 GSEventRunModal + 164
34  UIKitCore                     	       0x18d6f33d0 -[UIApplication _run] + 888
35  UIKitCore                     	       0x18d6f3034 UIApplicationMain + 340
36  Jetpack                       	       0x105cdcd54 main + 480
37  dyld                          	       0x1a987c960 start + 2528
```

</details>

The crash on WordPress is definitely related to Core Data, and it looks like it can be reproduced by triggering the export when WordPress is in an inactive/killed state. As for the crash on Jetpack, I haven't investigated further, but seeing that it's not crashing on your side (rather, the app enters an invalid state) makes me think it's related to the logout logic. Regardless, I think these are likely two separate issues, and we'll probably need to address them separately from this PR.

**Comment by twstokes** [comment]
> Regardless, I think these are likely two separate issues, and we'll probably need to address them separately from this PR.

That sounds good @dvdchr! Thanks for checking behind me! 🙇

**Comment by twstokes** [comment]
Synced to the release branch via 1467d0442c96093c53eb162bfdd1715a910bc149 to pull in #19792.

**Review by twstokes — APPROVED** [review]
Great job @dvdchr! 🚀 All tests passed for me. I only left one note about test #9 that's a future consideration.

File: WordPress/Classes/Utility/Migration/ContentMigrationCoordinator.swift
Comment (twstokes) [reply WordPress/Classes/Utility/Migration/ContentMigrationCoordinator.swift:72]: - WordPress check added in 42789939adc085145cd14416742b89845447bf21.
- Confirmed that `21.3` will be skipped. 👍

**Comment by spencertransier** [comment]
This fix made it into the `21.4.0.1` build.

**Pull request merged**
